app/src/main/java/io/homeassistant/companion/android/background/LocationBroadcastReceiver.kt

```diff
@@ -0,0 +1,144 @@
+package io.homeassistant.companion.android.background
+
+import android.Manifest.permission.ACCESS_COARSE_LOCATION
+import android.app.PendingIntent
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.content.pm.PackageManager
+import android.os.BatteryManager
+import android.os.Build
+import android.util.Log
+import androidx.core.app.ActivityCompat
+import com.google.android.gms.location.LocationRequest
+import com.google.android.gms.location.LocationResult
+import com.google.android.gms.location.LocationServices
+import io.homeassistant.companion.android.DaggerReceiverComponent
+import io.homeassistant.companion.android.common.dagger.Graph
+import io.homeassistant.companion.android.domain.integration.IntegrationUseCase
+import io.homeassistant.companion.android.domain.integration.UpdateLocation
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+class LocationBroadcastReceiver : BroadcastReceiver() {
+
+    companion object {
+        const val ACTION_REQUEST_LOCATION_UPDATES =
+            "io.homeassistant.companion.android.background.REQUEST_UPDATES"
+        const val ACTION_PROCESS_LOCATION =
+            "io.homeassistant.companion.android.background.PROCESS_UPDATES"
+
+        private const val TAG = "LocBroadcastReceiver"
+    }
+
+    @Inject
+    lateinit var integrationUseCase: IntegrationUseCase
+
+    private val mainScope: CoroutineScope = CoroutineScope(Dispatchers.Main + Job())
+
+    override fun onReceive(context: Context, intent: Intent) {
+        DaggerReceiverComponent.builder()
+            .appComponent(Graph(context).appComponent)
+            .build()
+            .inject(this)
+
+        when (intent.action) {
+            Intent.ACTION_BOOT_COMPLETED -> requestUpdates(context)
+            ACTION_REQUEST_LOCATION_UPDATES -> requestUpdates(context)
+            ACTION_PROCESS_LOCATION -> handleUpdate(context, intent)
+            else -> Log.w(TAG, "Unknown intent action: ${intent.action}!")
+        }
+    }
+
+    private fun requestUpdates(context: Context) {
+        Log.d(TAG, "Registering for location updates.")
+
+        if (ActivityCompat.checkSelfPermission(
+                context,
+                ACCESS_COARSE_LOCATION
+            ) != PackageManager.PERMISSION_GRANTED
+        ) {
+            Log.w(TAG, "Not starting location reporting because of permissions.")
```

Comment:
Maybe a toast message to tell the user that location reporting did not start because of missing permissions?

Reply:
And then also ask for the permissions 👍

Reply:
If we make it to this point the user declined to give the permissions, do we really want to ask every time we start the app if they don't want location updates?

Reply:
Oh, right, we should not.

Reply:
Android will offer to deny any further permission requests after the first one has been denied.
But it'd make more sense if there was an option in the app to toggle location reporting. When turned on, just ask for permission there.

Reply:
Agreed, but out side of the scope of this PR I would say though.  We still need a settings UI for basically everything (urls, device registration changes, location settings)